kerngen/pisa_generators/rescale.py

```diff
@@ -0,0 +1,139 @@
+        op_last_q: bool = False
+
+    @dataclass
+    class PartialOpPolys:
+        """Polynomials used in partial ops"""
+
+        output: Polys
+        input0: Polys
+        input1: Polys
+        input_remaining_rns: Polys
+
+    def partial_op(
+        self,
+        op,
+        polys: PartialOpPolys,
+        options: PartialOpOptions,
+    ):
+        """ "A helper function to perform partial operation, such as add/sub on last half (input1) to all of input0"""
+        last_q = self.input0.rns - 1
+        return [
+            op(
+                self.context.label,
+                polys.output(part, last_q if options.output_last_q else q, unit),
+                polys.input0(part, last_q if options.input0_last_q else q, unit),
+                polys.input1(
+                    0 if options.input1_first_part else part,
+                    last_q if options.input1_last_q else q,
+                    unit,
+                ),
+                last_q if options.op_last_q else q,
+            )
+            for part, q, unit in product(
+                range(polys.input_remaining_rns.parts),
+                range(polys.input_remaining_rns.rns),
+                range(self.context.units),
+            )
+        ]
+
+    def add_last_half(self, output, input0, input1, input_remaining_rns):
+        """Add input0 to input1 (first part)"""
+        return self.partial_op(
+            pisa_op_add,
+            self.PartialOpPolys(output, input0, input1, input_remaining_rns),
+            self.PartialOpOptions(
+                output_last_q=True,
+                input0_last_q=True,
+                input1_last_q=True,
+                input1_first_part=True,
+                op_last_q=True,
+            ),
+        )
+
+    def sub_last_half(self, output, input0, input1, input_remaining_rns):
+        """Subtract input1 (first part) with input0 (last RNS)"""
+        return self.partial_op(
+            pisa_op_sub,
+            self.PartialOpPolys(output, input0, input1, input_remaining_rns),
+            self.PartialOpOptions(input0_last_q=True, input1_first_part=True),
+        )
+
+    def to_pisa(self) -> list[PIsaOp]:
+        """Return the p-isa code to perform an mod switch down"""
+        # Convenience and Immediates
+        context = self.context
+        last_q = self.input0.rns - 1
+        one = Immediate(name="one")
+        r2 = Immediate(name="R2", rns=last_q)
+        iq = Immediate(name="iq", rns=last_q)
+        q_last_half = Polys("qLastHalf", 1, self.input0.rns)
+        q_i_last_half = Polys("qiLastHalf", 1, rns=last_q)
+
+        # Drop down input rns
+        input_last_rns = Polys.from_polys(self.input0, mode="last_rns")
+        input_remaining_rns = Polys.from_polys(self.input0, mode="drop_last_rns")
+
+        # Temp.
+        y = Polys(
+            "y",
+            input_last_rns.parts,
+            input_last_rns.rns,
+            start_rns=input_last_rns.start_rns,
+        )
```

Comment:
```suggestion
        y = Polys(
            name="y",
            parts=input_last_rns.parts,
            rns=input_last_rns.rns,
            start_parts=0,
            start_rns=input_last_rns.start_rns
        )
```
For consistency we should agree with how to pass the variables to a function. We either use positional or by name, but not mixed as it will lead to confusion to readers of the code. In this case, and in my opinion we should do by name, again for clarity.